pkg/engine/lifecycletest/source_query_test.go

```diff
@@ -0,0 +1,150 @@
+				) (plugin.ReadResult, resource.Status, error) {
+					return plugin.ReadResult{Outputs: resource.PropertyMap{}}, resource.StatusOK, nil
+				},
+			}, nil
+		}),
+	}
+
+	programF := deploytest.NewLanguageRuntimeF(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		_, _, _, err := monitor.RegisterResource(providers.MakeProviderType("pkgA"), "provA", true)
+		assert.Error(t, err)
+		assert.ErrorContains(t, err, "Query mode does not support creating, updating, or deleting resources")
+
+		_, _, _, err = monitor.RegisterResource("pkgA:m:typA", "resA", true, deploytest.ResourceOptions{})
+		assert.Error(t, err)
+		assert.ErrorContains(t, err, "Query mode does not support creating, updating, or deleting resources")
+
+		_, _, err = monitor.ReadResource("pkgA:m:typA", "resA", "read-id", "", nil, "", "", "")
+		assert.Error(t, err)
+		assert.ErrorContains(t, err, "Query mode does not support reading resources")
+		return nil
+	})
+
+	plugCtx, err := plugin.NewContext(
+		diagtest.LogSink(t), diagtest.LogSink(t),
+		deploytest.NewPluginHostF(nil, nil, programF, loaders...)(),
+		nil, "", nil, false, nil)
+	assert.NoError(t, err)
+
+	src, err := deploy.NewQuerySource(context.Background(), plugCtx, &deploytest.BackendClient{}, &deploy.EvalRunInfo{
+		Proj: &workspace.Project{
+			Name: "query-program",
+		},
+	}, nil, nil)
+	assert.NoError(t, err)
+	assert.NoError(t, src.Wait())
+}
+
+func TestRunQuery_call(t *testing.T) {
+	t.Parallel()
+
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{
+				InvokeF: func(
+					tok tokens.ModuleMember, inputs resource.PropertyMap,
+				) (resource.PropertyMap, []plugin.CheckFailure, error) {
+					name := inputs["name"]
+					ret := "unexpected"
+					if name.IsString() {
+						ret = "Hello, " + name.StringValue() + "!"
+					}
+
+					return resource.NewPropertyMapFromMap(map[string]interface{}{
+						"message": ret,
+					}), nil, nil
+				},
+				CallF: func(monitor *deploytest.ResourceMonitor, tok tokens.ModuleMember,
+					args resource.PropertyMap, info plugin.CallInfo, options plugin.CallOptions,
+				) (plugin.CallResult, error) {
+					ret := "unexpected"
+					if args["name"].IsString() {
+						ret = "Hello, " + args["name"].StringValue() + "!"
+					}
+
+					return plugin.CallResult{
+						Return: resource.NewPropertyMapFromMap(map[string]interface{}{
+							"message": ret,
+						}),
+					}, nil
+				},
+			}, nil
+		}),
+	}
+
+	programF := deploytest.NewLanguageRuntimeF(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		outs, _, _, err := monitor.Call("pkgA:m:typA/methodA", resource.PropertyMap{
+			"name": resource.NewStringProperty("bar"),
+		}, "", "")
+		assert.NoError(t, err)
+		assert.True(t, outs.DeepEquals(resource.PropertyMap{
+			"message": resource.NewStringProperty("Hello, bar!"),
+		}), "outs was %v", outs)
```

Comment:
I don't think you need `DeepEquals` here. Feedback applies elsewhere.

```suggestion
		assert.Equal(t, resource.PropertyMap{
			"message": resource.NewStringProperty("Hello, bar!"),
		}, outs)
```